### PR TITLE
fix_: sync fallback notification

### PR DESCRIPTION
**Pull request by qfrank**
This PR already contains the changes made in [PR 5869](https://github.com/status-im/status-go/pull/5869), because these two PRs have some intersections. To make things simple, let's just use this PR instead.

This PR mainly focuses on solving mobile [issue 21260](https://github.com/status-im/status-mobile/issues/21260)
We should not create/delete the notification if the target is not equal current installation id.

Relate mobile [PR 21315](https://github.com/status-im/status-mobile/pull/21315)

status: ready.

### CODE REVIEW

File: cmd/statusd/server/endpoints.go
Comment (jrainville) [inline cmd/statusd/server/endpoints.go:1]: Are these changes needed? Are they auto-generated?

File: protocol/messenger.go [hunk 1]
```diff
@@ -3150,6 +3150,8 @@ type ReceivedMessageState struct {
 	AllInstallations *installationMap
 	// List of communities modified
```
Comment (jrainville) [inline protocol/messenger.go]: Unrelated, but this comment doesn't fit the property. It's probably my fault

File: protocol/messenger_handler.go [hunk 2]
```diff
@@ -1321,6 +1321,8 @@ func (m *Messenger) HandleSyncPairInstallation(state *ReceivedMessageState, mess
 	// TODO(samyoul) remove storing of an updated reference pointer?
 	state.AllInstallations.Store(message.InstallationId, installation)
 	state.ModifiedInstallations.Store(message.InstallationId, true)
+	targeted := message.TargetInstallationId == m.installationID
+	state.ModifiedInstallationsTargetedToThisDevice.Store(message.InstallationId, targeted)
```
Comment (jrainville) [inline protocol/messenger_handler.go]: Can we only store the `InstallationId` and not `targeted`. The name of the property makes it seem like it would only contain targeted installations already.

**Review by jrainville — APPROVED** [review]
Looks good, just some small comments

**Review by jrainville — CHANGES_REQUESTED** [review]
Changing my review because I just tested this branch and I no longer get any AC notification on both my apps. 

Unless I need to modify my code in the Desktop app, but I doubt it since you just changed the way the AC notifs are triggered.

Here is an example, I synced the left app to the right one, so notifs of both:
image

**Comment by qfrank** [comment]
> because I just tested this branch and I no longer get any AC notification on both my apps.

thanks for reviewing! that's weird, you can see mobile QA test [result](https://github.com/status-im/status-mobile/pull/21315), also the tests in `messenger_pairing_and_syncing_test.go` works, also I can receive the AC notifications(ActivityCenterNotificationTypeNewInstallationReceived/ActivityCenterNotificationTypeNewInstallationCreated) when testing on local mobile simulators, is there any chance that it's because your network issue or the message reliability issue? could you double check? 🤔 @jrainville

File: protocol/messenger_handler.go (hunk 2 above)
Comment (qfrank) [reply protocol/messenger_handler.go]: what about new name: `ModifiedInstallationsTargetedToThisDevice ` -> `TargetedInstallations` ?

File: protocol/messenger.go (hunk 1 above)
Comment (qfrank) [reply protocol/messenger.go]: Removed the comment. I think the name of the variable shows it clearly.

File: cmd/statusd/server/endpoints.go
Comment (qfrank) [reply cmd/statusd/server/endpoints.go:1]: updated by make lint-fix, reverted

**Comment by jrainville** [comment]
> > because I just tested this branch and I no longer get any AC notification on both my apps.
> 
> thanks for reviewing! that's weird, you can see mobile QA test [result](https://github.com/status-im/status-mobile/pull/21315), also the tests in `messenger_pairing_and_syncing_test.go` works, also I can receive the AC notifications(ActivityCenterNotificationTypeNewInstallationReceived/ActivityCenterNotificationTypeNewInstallationCreated) when testing on local mobile simulators, is there any chance that it's because your network issue or the message reliability issue? could you double check? 🤔 @jrainville

Ah! I know what happened. When I tested your branch, I still had an  issue in my own branch.

I'll test your branch again now that I fixed my own issue.

**Review by jrainville — APPROVED** [review]
All good. I retested with my fixed branch and your changes work fine.

I also tested by re-importing a new one just with the seed phrase and it didn't get the AC notification :+1:

File: protocol/requests/enable_installation_and_pair.go [hunk 3]
```diff
@@ -8,6 +8,9 @@ var ErrEnableInstallationAndPairInvalidID = errors.New("enable installation and
 
 type EnableInstallationAndPair struct {
 	InstallationID string `json:"installationId"`
+
+	// used for testing
+	getInstallationID func() string
```
Comment (igor-sirotin) [inline protocol/requests/enable_installation_and_pair.go]: Although I like the simplicity, I don't think it's a good approach. Production code should not contain test things.

I think it should be feasible to unit-test `saveDataAndPrepareResponse` directly, by passing request with specific `installationID`, can you please try?

File: protocol/protobuf/pairing.proto [hunk 4]
```diff
@@ -91,8 +91,9 @@ message SyncPairInstallation {
   string installation_id = 2;
   string device_type = 3;
   string name = 4;
-  // following fields used for local pairing
+  // used for local pairing
   uint32 version = 5;
+  string target_installation_id = 6;
```
Comment (igor-sirotin) [inline protocol/protobuf/pairing.proto:96]: Actually, it's weird that we have this here. I'd rather `SyncInstallation` message and a `PairInstallation` that would compose the former one. But it might be too difficult to refactor now I guess.

File: protocol/messenger_pairing_and_syncing_test.go [hunk 5]
```diff
@@ -58,12 +59,111 @@ func (s *MessengerPairingSuite) TestEnableNonExistingInstallation() {
...
+}
+
+func (s *MessengerPairingSuite) expectInstallationAndACNotificationNewInstallationReceived(m *Messenger, installationID string) {
+	_, err := WaitOnMessengerResponse(
+		m,
+		func(r *MessengerResponse) bool {
+			for _, i := range r.Installations() {
+				if i.ID == installationID && len(r.ActivityCenterNotifications()) == 1 && r.ActivityCenterNotifications()[0].Type == ActivityCenterNotificationTypeNewInstallationReceived {
+					return true
+				}
```
Comment (igor-sirotin) [inline protocol/messenger_pairing_and_syncing_test.go]: We shouldn't wait for AC notification in the loop.
Instead, just wait for the installation and check the AC after `WaitOnMessengerResponse`.

File: protocol/messenger_pairing_and_syncing_test.go [hunk 6]
```diff
@@ -58,12 +59,111 @@ func (s *MessengerPairingSuite) TestEnableNonExistingInstallation() {
...
+	)
+	s.Require().NoError(err)
+
+	ac, err := alice3.ActivityCenterNotification(types.FromHex(alice1.installationID))
+	s.Require().NoError(err)
+	s.Require().Equal(ActivityCenterNotificationTypeNewInstallationCreated, ac.Type)
+	s.Require().False(ac.Deleted)
+}
+
+func (s *MessengerPairingSuite) expectInstallationAndACNotificationNewInstallationReceived(m *Messenger, installationID string) {
```
Comment (igor-sirotin) [inline protocol/messenger_pairing_and_syncing_test.go]: I understand your intention, but this is a waaaaay too long function name 😄 
`expectInstallationReceived` is enough I think.

File: protocol/messenger_pairing_and_syncing_test.go (hunk 5 above)
Comment (igor-sirotin) [reply protocol/messenger_pairing_and_syncing_test.go]: Same thing in other `WaitOnMessengerResponse` in this PR

File: protocol/requests/enable_installation_and_pair.go (hunk 3 above)
Comment (igor-sirotin) [reply protocol/requests/enable_installation_and_pair.go]: If you find it not possible to unit-test, then we should go the proper mock way.
1. Keep `EnableInstallationAndPair` as before, no custom callback properties
2. Define `EnableInstallationAndPairInterface` interface with `GetInstallationID` method
3. Implement `func (r *EnableInstallationAndPair) GetInstallationID()`
4. Change `EnableInstallationAndPair` signature to take `EnableInstallationAndPairInterface`.
5. Use `gomock` to generate a mock in tests
6. Use the generated mocks in test

Comment (qfrank) [reply protocol/requests/enable_installation_and_pair.go]: >  Production code should not contain test things.

Make sense!

>  unit-test saveDataAndPrepareResponse directly

I don't think it's a good idea to write unit test for `saveDataAndPrepareResponse`, `saveDataAndPrepareResponse` is a bit complicated and composed, it's not that `pure`, unit test and integration test have different purpose and different usage scenarios

File: protocol/protobuf/pairing.proto (hunk 4 above)
Comment (qfrank) [reply protocol/protobuf/pairing.proto:96]: 💔

File: protocol/messenger_pairing_and_syncing_test.go (hunk 5 above)
Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: > just wait for the installation and check the AC after WaitOnMessengerResponse

Sometimes we don't want the `merged response`, I don't have a strong opinion on how we handle this really , WDYT? 🙂

File: protocol/messenger_pairing_and_syncing_test.go (hunk 6 above)
Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: renamed

File: protocol/requests/enable_installation_and_pair.go (hunk 3 above)
Comment (qfrank) [reply protocol/requests/enable_installation_and_pair.go]: > If you find it not possible to unit-test, then we should go the proper mock way.
> 
> 1. Keep `EnableInstallationAndPair` as before, no custom callback properties
> 2. Define `EnableInstallationAndPairInterface` interface with `GetInstallationID` method
> 3. Implement `func (r *EnableInstallationAndPair) GetInstallationID()`
> 4. Change `EnableInstallationAndPair` signature to take `EnableInstallationAndPairInterface`.
> 5. Use `gomock` to generate a mock in tests
> 6. Use the generated mocks in test

created interface `InstallationIDProvider`

Comment (igor-sirotin) [reply protocol/requests/enable_installation_and_pair.go]: > _`saveDataAndPrepareResponse` is a bit complicated and composed_

On the opposite, this is a very good reason for unit-testing 🙂

File: protocol/messenger_pairing_and_syncing_test.go (hunk 5 above)
Comment (igor-sirotin) [reply protocol/messenger_pairing_and_syncing_test.go]: > _Sometimes we don't want the merged response_

@qfrank 

1. Why?
2. Now you're checking the merged response anyway

Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: > 1. Why?

e.g. some special type(bool,string...) properties not support merging, but that's another story.

> 2. Now you're checking the merged response anyway

Oh you're right, some mistakes happened for my memory 😅 ... I meant use `tt.RetryWithBackOff` directly for special types that not support merging

File: protocol/requests/enable_installation_and_pair.go (hunk 3 above)
Comment (qfrank) [reply protocol/requests/enable_installation_and_pair.go]: I think it's easier for writing integration test as there already exist some cases.

File: protocol/messenger_pairing_and_syncing_test.go (hunk 5 above)
Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: fixed

File: protocol/requests/enable_installation_and_pair.go (hunk 3 above)
Comment (igor-sirotin) [reply protocol/requests/enable_installation_and_pair.go]: Well I didn't say it's gonna be easier 😄 
But in the future we should try hard to write more unit tests than integration tests.

File: protocol/requests/enable_installation_and_pair.go [hunk 7]
```diff
@@ -6,6 +6,11 @@ import (
 
 var ErrEnableInstallationAndPairInvalidID = errors.New("enable installation and pair: invalid installation id")
 
+type InstallationIDProvider interface {
+	GetInstallationID() string
+	Validate() error
+}
+
```
Comment (igor-sirotin) [inline protocol/requests/enable_installation_and_pair.go]: In Go interfaces are usually defined in place of usage, so in this case somewhere next to `SendPairInstallation`.
Because `InstallationIDProvider` doesn't make any sense in the `requests` dir

File: protocol/messenger_pairing_and_syncing_test.go [hunk 8]
```diff
@@ -58,12 +59,127 @@ func (s *MessengerPairingSuite) TestEnableNonExistingInstallation() {
...
+	if !m.getInstallationIDInvokedFirst {
+		m.getInstallationIDInvokedFirst = true
+		return m.installationID
+	}
+	return "wrong installation ID"
+}
+
+func (m *mockEnableInstallationAndPair) Validate() error {
+	return nil
+}
```
Comment (igor-sirotin) [inline protocol/messenger_pairing_and_syncing_test.go]: Why not generate a mock instead? 🤔

**Review by igor-sirotin — APPROVED** [review]
Thanks for the addressed comments!

File: protocol/messenger_pairing_and_syncing_test.go (hunk 8 above)
Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: Because it's simple doing this without knowing mock, I won't learn it until I need depend on it heavily

File: protocol/requests/enable_installation_and_pair.go (hunk 3 above)
Comment (qfrank) [reply protocol/requests/enable_installation_and_pair.go]: > But in the future we should try hard to write more unit tests than integration tests.

is it because the long time we need to wait for test result? if yes, I feel a little sad 😭 Maybe we should try other solution

File: protocol/requests/enable_installation_and_pair.go (hunk 7 above)
Comment (qfrank) [reply protocol/requests/enable_installation_and_pair.go]: moved

File: protocol/messenger_pairing_and_syncing_test.go (hunk 8 above)
Comment (ilmotta) [reply protocol/messenger_pairing_and_syncing_test.go]: Nitpick: technically I think this is more like a stub than a mock. Mocks have other abilities built in, like tracking call count, tracking arguments, and in general other behavioral checks, like auto-implementing functions to return nil by default.

Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: so `stubEnableInstallationAndPair` should be a better name 🤔? @ilmotta

Comment (ilmotta) [reply protocol/messenger_pairing_and_syncing_test.go]: I would say so @qfrank. With a mock we usually write the expectations even before the test runs, saying for example, how many times a specific call is expected to happen, or dynamically making the mock respond to certain input. If we implement an interface and provide static output like what this PR does, then it's the perfect definition of a stub.

Sometimes it's hard to differentiate and name different test doubles. Sometimes it's a gray area.

Comment (qfrank) [reply protocol/messenger_pairing_and_syncing_test.go]: thank you, renamed
